Update ELLocation LocationManager to improve thread safety

nonsensery opened the pull request:
#### What does this PR do?

Changes some of the design of the internal `ELLocation.LocationManager` so that the data flow is more clear, and improves thread safety.
#### Any background context you want to provide?

This is a bigger change than I had initially intended to make. I started looking into fixing a thread-safety issue in `deregisterListener`:

``` swift
    func deregisterListener(listener: AnyObject) {
        if let theIndex = indexOfLocationListenerForListener(listener) {
            removeLocationListenerAtIndex(theIndex)
        }
    }
```

There was no `synchronized` call in this method (but there were calls in `indexOf…` and `removeLocationListener…`). Effectively, the code worked like this:
1. Acquire lock.
2. Find index of listener.
3. Release lock.
4. Reacquire lock.
5. Remove listener using index from step (2)
6. Release lock.

Between steps (3) and (4) the listener array could have been modified, invalidating the index.

As I tried to implement that, I realized that there were several places where state changes were originating, which made them difficult to isolate and synchronize.

I ended up redesigning the internals of `ELLocation.LocationManager` so that changes to the internal state go through one central "reducer" that then updates the single internal `CLLocationManager`'s state. All access to internal state is `synchronized` to make it thread safe.

The "reducer" data flow works like this:

Inputs:
- Array of listeners with `accuracy` and `updateFrequency`.
- Core location authorization status.

Intermediates:
- Maximum accuracy from the listeners.
- Maximum update frequency from the listeners.

Outputs:
- Desired accuracy for the underlying location manager.
- Distance filter for the location manager.
- Location updates on/off.
- Significant change monitoring on/off.
#### Where should the reviewer start?
- Most of the changes are to the `LocationManager` class.
- I also added a new private enum for the two different location monitoring modes ("significant updates" and "standard").
- Most of the state reduction happens through series of computed properties. For example, the `coreLocationDesireAccuracy` is computed from the `accuracy`, which is computed from `allLocationListeners`.
#### How should this be manually tested?

There is a sample app in the project. If you run the sample app using the old version and this new version, they should behave the same.

nonsensery committed:
Change LocationManager to use a more one-way data flow

nonsensery committed:
OMG ... fix the code to actually turn on location updates!

bsneed commented:
@samgrover ^

steveriggins commented:
Unit tests possible for the code that is not location dependent, such as adding/removing listeners?

nonsensery commented:
@steveriggins I will look into adding unit tests. I think it should be reasonably easy.

nonsensery committed:
Fix goofy code that updates monitoring mode

nonsensery committed:
Add private less-than operators to clarify code

This hides away the `rawValue` so it doesn’t leak out into the code
that consumes the enums.

nonsensery committed:
Switch to non-singleton instance for unit tests

nonsensery committed:
Don't add same handler multiple times in tests

This behavior is currently allowed, but it’s not a good idea and should
probably be explicitly forbidden in the future.

nonsensery committed:
Remove expectations for no-listener behavior

This behavior is not defined

nonsensery committed:
Add simple tests for CL desired accuracy

nonsensery committed:
Add basic test to verify that listener is notified

samgrover commented:
@nonsensery Thanks for this PR and for taking care of the feedback I gave yesterday. All helpful and good stuff!
